src/isolation/tempdir.rs

```diff
@@ -1,25 +1,132 @@
-use std::{fs::Permissions, os::unix::fs::PermissionsExt, path::PathBuf};
+use std::{
+	fs::{Permissions, create_dir, set_permissions},
+	os::unix::fs::{MetadataExt, PermissionsExt},
+	path::PathBuf,
+};
 
 use tempfile::{Builder, TempDir, env};
 use uuid::Uuid;
 
+/// Gets own user ID to append to subdirectory (e.g. `/tmp/uhyve-2`).
+///
+/// Defaults to 0 if accessing own user ID is not possible and hopes
+/// for the best.
+fn get_uid() -> u32 {
+	// Use user ID to create subdirectory.
+	std::fs::metadata("/proc/self")
+		.map(|m| m.uid())
+		.unwrap_or_else(|e| {
+			error!(
+				"Unable to get own user ID to generate temporary directory, defaulting to 0: {e}"
+			);
+			0
+		})
+}
+
 /// Creates a temporary directory.
 ///
+/// To keep the temporary directory clean and because we need to include the
+/// parent directory in [`crate::isolation::landlock::initialize`], a sub-folder
+/// called `uhyve-{uid}` (e.g. `/tmp/uhyve-2`) is used to create temporary
+/// directories under. The user ID is used to limit access to the user running Uhyve,
+/// while not preventing other users from running Uhyve instances themselves.
+///
+/// After creating a directory, the permissions are set to 0o700 (read, write, execute
+/// for the user). Those permissions are checked during the runtime.
+///
 /// * `dir_path` - The location in which the temporary directory should be created.
 pub fn create_temp_dir(dir_path: &Option<String>) -> TempDir {
 	let dir = Builder::new()
 		.permissions(Permissions::from_mode(0o700))
 		.prefix("uhyve-")
 		.suffix(&Uuid::new_v4().to_string())
-		.tempdir_in(
-			dir_path
-				.as_ref()
-				.map(PathBuf::from)
-				.unwrap_or_else(env::temp_dir),
-		)
-		.expect("The temporary directory could not be created.");
+		.tempdir_in(dir_path.as_ref().map(PathBuf::from).unwrap_or_else(|| {
+			let mut env_temp_dir = env::temp_dir();
+			// Sane environments presumed to have world-writable temporary directory.
+			debug_assert!(!env_temp_dir.metadata().unwrap().permissions().readonly());
+
+			env_temp_dir.push(format!("uhyve-{}", get_uid()));
 
-	assert!(!dir.path().metadata().unwrap().permissions().readonly());
+			// If the subfolder doesn't exist, try to create it.
+			// If that isn't possible, panic.
+			if env_temp_dir.metadata().is_err() {
+				create_dir(&env_temp_dir).unwrap_or_else(|e| {
```

Comment:
Consider using unix-specific https://doc.rust-lang.org/std/fs/struct.DirBuilder.html#method.mode `std::fs::DirBuilder` here + `std::os::unix::fs::DirBuilderExt::mode` to avoid the delay between creation of directory and setting its mode.